### ISSUE TITLE
Support semi-linear merge option for pull requests

### ISSUE THREAD

**Issue by waldyrious** [isaacs/github#1017]
This would be equivalent to rebasing a PR without merging it (tracked as #88, most explicitly described in @jagthedrummer's [comment](https://github.com/isaacs/github/issues/88#issuecomment-266773758)), and then using the normal merge option (which always creates a merge commit, even if the merge could be fast-forwarded).

Such a workflow can be implemented via the command line, or using third-party tools, as @galulex illustrated [here](https://github.com/isaacs/github/issues/88#issuecomment-286639281), but it would be best if it was a natively supported additional option in GitHub's web UI.

FYI, [GitLab has implemented this](https://gitlab.com/gitlab-org/gitlab-ee/merge_requests/110) about a year ago, and it's a [popular request for Bitbucket](https://bitbucket.org/site/master/issues/6106) as well.

To make this more concrete and visual, I found that the diagram from the post [A tidy, linear Git history](http://www.bitsnbites.eu/a-tidy-linear-git-history/) explains the difference between these workflows quite well in a visual way, and I took the liberty to add a pure linear workflow to the diagram to quickly convey the differences between `merge` vs. `rebase && merge --ff-only` (linear) vs. `rebase && merge --no-ff` (semi-linear):

![git history diagrams](https://gitlab.com/gitlab-org/gitlab-ee/uploads/eea1c55b2bb1fdca939d79b93f65dcd6/merge_vs._rebase-ff_vs._rebase-no-ff.png)

**Comment by insearching** [comment]
Any updates here?

**Comment by mbitsnbites** [comment]
I second this suggestion. Most of my projects follow the [stable mainline](https://www.bitsnbites.eu/a-stable-mainline-branching-model-for-git/) model that relies on `rebase && merge --no-ff` (a.k.a. "merge commit with semi-linear history").

![image](https://user-images.githubusercontent.com/1193992/95950684-50485b80-0df5-11eb-862f-3a2d8eccf1d0.png)

Right now I have to go through hoops to achieve the same thing in GitHub:
1. Check if the PR branch is based on the latest commit on master. This is quite cumbersome in GitHub since that information is not available in the PR (to my knowledge), but rather I have to go to the source/fork repo & branch and check the history to find the SHA of the commit that comes before the first PR commit, and compare that to the SHA of the tip of master in the target repo & branch.
2. If the PR branch is based on the tip of master, great! Just pick the "Create merge commit" option and merge. Done! *Obviously this assumes that no commits are pushed to the master branch after the SHA comparison was done. In personal projects with a single maintainer that is usually not a problem. In larger projects this can be a real problem though.*
3. Otherwise I have two options:
  a. Instruct the PR author to rebase the branch & force push. This can be problematic if there are several concurrent PR:s in progress (and honestly it can be a bit awkward to put this extra work on the contributor).
  b. Rebase the branch myself, close the old PR, create a new PR and use the "Create merge commit" option.

In any event, this means a lot of manual work that could and should be automated.

Ideally, any CI checks should be re-run with the rebased branch before doing `merge --no-ff`, but even without that functionality there would be great value in a fourth "Create merge commit with semi-linear history" option.

**Comment by mbitsnbites** [comment]
The extra option could look something like this:
![image](https://user-images.githubusercontent.com/1193992/95956876-37dd3e80-0dff-11eb-93c7-970361601c9a.png)

**Comment by K3UL** [comment]
Would that include the fact to preserve commits SHAs ? Github rebasing and changing the SHA is a problem in my case to compare where a branch is ahead of another

**Comment by waldyrious** [comment]
@K3UL The only way to keep the graph semi-linear (i.e. avoid crossing lines as in the first diagram in the opening comment) would be to rebase PR branches whose base has lagged behind the target branch, and that would necessarily change the commit hashes. But I would definitely be in favor of preserving the branch's existing commit hashes if it doesn't require rebasing (i.e. if a fast-forward merge would be possible).

**Comment by marcus-blaettermann-ventoro** [comment]
@waldyrious there already is a protected branch rule _"Require branches to be up to date before merging"_ that enforces the PR commits to be ahead of the current master — thus already be rebased. Enabling this rule could be a requirement to be able to use merging with linear history.

I think the big advantage here is: the commit being build in the PR is guaranteed to be the one that ends up being the new `HEAD` of master. Which is great, because there is absolutely nothing changing after the PR is closed. We already enforce this in our repos, but currently have to do this manually and merge the feature-branch by hand on our locale machine.

It might be a bit cumbersome for large repos, with several PRs being open in parallel, but I think for a lot of repos, the super clean history is worth it.

**Comment by essenmitsosse** [comment]
@waldyrious there already is a protected branch rule "Require branches to be up to date before merging" that enforces the PR commits to be ahead of the current master — thus already be rebased. Enabling this rule could be a requirement to be able to use merging with linear history.

I think the big advantage here is: the commit being build in the PR is guaranteed to be the one that ends up being the new HEAD of master. Which is great, because there is absolutely nothing changing after the PR is closed. We already enforce this in our repos, but currently have to do this manually and merge the feature-branch by hand on our locale machine.

It might be a bit cumbersome for large repos, with several PRs being open in parallel, but I think for a lot of repos, the super clean history is worth it.

**Comment by chucklu** [comment]
@waldyrious  The problem of fast-forward merge is, you could not easily tell a merge operation from the log graph. That's why we need semi-linear merge.  Semi-linear will not re-generate new commits for existing commits, when the commits can be fast-forwarded, it will only generate a new merge commit.

**Comment by sandstrom** [comment]
If you want this feature, send it to the Github team working on PRs, merge button, etc.

Take 1 minute and send your feedback here:
https://support.github.com/contact/feedback?category=prs-and-code-review

**Comment by waldobronchart** [comment]
For anyone watching this thread, I wrote a script to do this until Github adds this option.

Get the script here: https://pypi.org/project/git-pr-linear-merge/

It does what the original post describes: rebase, then merge without fast-forward. 

Documentation can be found on the project page.

**Comment by black-puppydog** [comment]
@waldobronchart the GH repo linked from PyPI is private, you might wanna change that. :)
Also why did you use "linear" for what is called "semi-linear" here?

**Comment by waldobronchart** [comment]
I gotta fix the source link, the repo is https://github.com/waldobronchart/git-pr-linear-merge

> Also why did you use "linear" for what is called "semi-linear" here?

Yeah I realized it after that that was the standard name, but I don't want to rename the package now because it'll cause confusion 🤷
